src/Signal-Estimator-GUI/mainwindow.cpp

```diff
@@ -0,0 +1,243 @@
+        if (std::strstr(buffer,"card") != NULL && std::strstr(buffer,"device") != NULL) // if line has both card and device in it
+            this->ui->OutputDevices->addItem(QString::fromStdString(buffer)); // add to combobox
+    }
+
+    std::unique_ptr<FILE,decltype(&pclose)> pipe2(popen("arecord -l","r"),pclose);
+    if(!pipe2){
+        throw std::runtime_error("Failed to run arecord");
+    }
+
+    while (fgets(buffer,128,pipe2.get()) != nullptr){
+        if (std::strstr(buffer,"card") != NULL && std::strstr(buffer,"device") != NULL) // if line has both card and device in it
+            this->ui->InputDevices->addItem(QString::fromStdString(buffer)); // add to combobox
+    }
+    this->show();
+}
+
+MainWindow::~MainWindow()
+{
+    delete ui;
+}
+
+void MainWindow::update_graphs(){ // update both input and output signal graphs
+    //this->ui->InputSig->detachItems(QwtPlotItem::Rtti_PlotCurve,false);
+
+    this->ui->InputSig->updateAxes();
+    this->ui->InputSig->replot();
+
+    //this->ui->OutputSig->detachItems(QwtPlotItem::Rtti_PlotCurve,false);
+
+    this->ui->OutputSig->updateAxes();
+    this->ui->OutputSig->replot();
+    QApplication::processEvents();
+}
+
+
+
+void MainWindow::read_graph_data(const char* cmd, QwtPlotCurve* incurve, QwtPlotCurve* outcurve,
+                                 QVector<double>& inx, QVector<double>& iny,
+                                 QVector<double>& outx, QVector<double>& outy, std::mutex& m){
+     char buffer[128];
+     char* token;
+     int counter = 0;
+
+     // run pipe with signal_estimator
+     // This assumes that the signal_estimator is outputting to stdout with the type of signal prepended to the data values
+     // The only changes to the signal_estimator source are the fact that frame.hpp now has a function to get iotype
+     // and the file dumper dumps the data prepended with iotype to stdout if stdout is used as file dumper option.
+
+     std::unique_ptr<FILE,decltype(&pclose)> pipe2(popen(cmd,"r"),pclose);
+     if(!pipe2){
+         throw std::runtime_error("Failed to run signal_estimator");
+     }
+     // loop to get a line from the pipe
+
+     while (fgets(buffer,128,pipe2.get()) != nullptr && this->get_update_plots() == true){
+
+        token = std::strtok(buffer," \n"); // get iotype
+        m.lock();
+        // if iotype is input
+        if (strcmp(token, "Input") == 0){
+            token = std::strtok(NULL, " \n");
+            inx.append(std::stod(token)/1000000); // append to Qvector x-axis for input graph
+            token = std::strtok(NULL, " \n");
+            if (token == NULL) // if corresponding y cannot be read, toss back the x out
+                inx.pop_back();
+            else // append to Qvector y-axis for input graph
+                iny.append(std::stod(token));
+            if (counter % 20 == 0){ // set the data for QWT plot curve in the input graph
+                incurve->setSamples(inx, iny);
```

Comment:
I suggest to extract parsing of signal-estimator output to a function.

Rationale: separate UI from parsing and keep boths part cleaner.

For example it can be something like:

```
struct Point {
   PointType type; // input or output
   double timestamp;
   double value;
};

std::optional<Point> parseLine(const char*);
```

Or maybe:

```
std::optional<Point> readAndParseLine(QProcess); // may be QIODevice instead of QProcess
```